pyccel/utilities/metaclasses.py

```diff
@@ -4,23 +4,47 @@
 #------------------------------------------------------------------------------------------#
 """ Module containing metaclasses which are useful for the rest of pyccel
 """
-
+from inspect import signature
 
 __all__ = (
     'Singleton',
     'ArgumentSingleton',
 )
 
+
 class ArgumentSingleton(type):
-    """ Indicates that there is only one instance of the class
-    for any given set of arguments
+    """
+    Metaclass indicating that there is only one instance of the parametrised class.
+
+    Metaclass indicating that there is only one instance of the class for any given
+    set of arguments.
+
+    Parameters
+    ----------
+    name : str
+        The name of the class.
+    bases : tuple[class,...]
+        A tuple of the superclasses of the class.
+    dct : dict
+        A dictionary of the class attributes.
     """
     _instances = {}
```

Comment:
I find it a bit confusing to give an `_instances` attribute to the metaclass. This will mean that the dictionary is in the metaclass rather than in each of the actual argument-singleton types. Why don't we create the dictionary in the `__init__` instead?

Reply:
I agree. I just kept the old implementation without thinking about it. Fixed: 4c0369d6

Reply:
If the dictionary is local to each argument-singleton type, it sounds like there is no need to include `cls` in the dictionary key. What do you think?

Reply:
If my last comment is correct, then we have no need to use a dictionary in the "simple" singleton case. And in fact it would be beneficial not to have the metaclass `Singleton` subclass `ArgumentSingleton`, in order to simplify the code.

Reply:
I agree. I have removed it: b273ef98

The original reason for making `Singleton` subclass `ArgumentSingleton` was to allow a class (`DataType`) to have the general metaclass (`ArgumentSingleton`), and the subclasses to redefine it (which is only allowed if the new metaclass is a subclass). But most of the time we want `Singleton` anyway and I am not sure that this is really helpful. I have split them back up again. [3405e0f](https://github.com/pyccel/pyccel/pull/1760/commits/3405e0f80e036b27426e95718aa3c2b1208428b4)